src/Authentication/Authentication/Helpers/HttpMessageFormatter.cs

```diff
@@ -0,0 +1,359 @@
+      Headers.ContentType = new MediaTypeHeaderValue(DefaultMediaType);
+      Headers.ContentType.Parameters.Add(new NameValueHeaderValue(MsgTypeParameter, DefaultResponseMsgType));
+
+      InitializeStreamTask();
+    }
+
+    private HttpContent Content
+    {
+      get { return HttpRequestMessage != null ? HttpRequestMessage.Content : HttpResponseMessage.Content; }
+    }
+
+    /// <summary>
+    /// Gets the HTTP request message.
+    /// </summary>
+    public HttpRequestMessage HttpRequestMessage { get; private set; }
+
+    /// <summary>
+    /// Gets the HTTP response message.
+    /// </summary>
+    public HttpResponseMessage HttpResponseMessage { get; private set; }
+
+    private void InitializeStreamTask()
+    {
+      _streamTask = new Lazy<Task<Stream>>(() => Content?.ReadAsStreamAsync());
+    }
+
+    /// <summary>
+    /// Validates whether the content contains an HTTP Request or an HTTP Response.
+    /// </summary>
+    /// <param name="content">The content to validate.</param>
+    /// <param name="isRequest">if set to <c>true</c> if the content is either an HTTP Request or an HTTP Response.</param>
+    /// <param name="throwOnError">Indicates whether validation failure should result in an <see cref="Exception"/> or not.</param>
+    /// <returns><c>true</c> if content is either an HTTP Request or an HTTP Response</returns>
+    internal static bool ValidateHttpMessageContent(HttpContent content, bool isRequest, bool throwOnError)
+    {
+      if (content == null)
+      {
+        throw new ArgumentNullException("content");
+      }
+
+      MediaTypeHeaderValue contentType = content.Headers.ContentType;
+      if (contentType != null)
+      {
+        if (!contentType.MediaType.Equals(DefaultMediaType, StringComparison.OrdinalIgnoreCase))
+        {
+          if (throwOnError)
+          {
+            throw new ArgumentException("HttpMessageInvalidMediaType", "content");
+          }
+          else
+          {
+            return false;
+          }
+        }
+
+        foreach (NameValueHeaderValue parameter in contentType.Parameters)
+        {
+          if (parameter.Name.Equals(MsgTypeParameter, StringComparison.OrdinalIgnoreCase))
+          {
+            string msgType = UnquoteToken(parameter.Value);
+            if (!msgType.Equals(isRequest ? DefaultRequestMsgType : DefaultResponseMsgType, StringComparison.OrdinalIgnoreCase))
+            {
+              if (throwOnError)
+              {
+                throw new ArgumentException("HttpMessageInvalidMediaType", "content");
+              }
+              else
+              {
+                return false;
+              }
+            }
+
+            return true;
+          }
+        }
+      }
+
+      if (throwOnError)
+      {
+        throw new ArgumentException("HttpMessageInvalidMediaType", "content");
+      }
+      else
+      {
+        return false;
+      }
+    }
+
+
+    public static string UnquoteToken(string token)
```

Comment:
If the input string guaranteed **not** to have repeated quotes at the beginning and at the end, the whole function can be replaced with `token.Trim('"')`